keps/sig-network/3015-prefer-same-node/README.md

```diff
@@ -0,0 +1,525 @@
+they would be very happy with a way to _just_ express "prefer same
+zone" without worrying that this might thwart some theoretical future
+proxy. (And of course, if it _does_ thwart some theoretical future
+proxy, they could just change the Service definition at that point.)
+
+Also, while these alternative traffic distribution modes risk
+overloading endpoints if used poorly, the API does not need to assume
+that the service proxy is the _only_ component in the system that
+cares about balancing endpoint load; the user may be using other APIs
+such as Pod Topology Spread Constraints or Horizontal Autoscaling to
+ensure that load is handled appropriately, and in those scenarios, the
+alternatively traffic distribution modes may be completely safe.
+
+## Design Details
+
+### Renaming/Deprecation of `PreferClose`
+
+KEP-4444 defines `PreferClose` as:
+
+> * `PreferClose`: Indicates a preference for routing traffic to endpoints that
+>   are topologically proximate to the client. The interpretation of
+>   "topologically proximate" may vary across implementations and could encompass
+>   endpoints within the same node, rack, zone, or even region.
+
+We will add a new `TrafficDistribution` value, `PreferSameZone`, which
+is defined as follows:
+
+* `PreferSameZone`: Indicates a preference for routing traffic to
+  endpoints that are in the same zone as the client. In general, the
+  proxy should always route to a same-zone endpoint if any is
+  available.
+
+And we will say that henceforth, `PreferClose` is just a deprecated
+alias for `PreferSameZone`, and never means anything other than
+"prefer same zone".
+
+Note that this does not change the existing implementation of
+`PreferClose` at all, it only removes the possibility of future
+changes.
+
+(In theory, service proxies other than kube-proxy may already have
+been implementing `PreferClose` with semantics other than the ones
+kube-proxy used. Those implementations will need to be updated when
+they add support for this KEP.)
+
+### Addition of `PreferSameNode`
+
+We will add a new `TrafficDistribution` value, `PreferSameNode`, which
+is defined as follows:
+
+* `PreferSameNode`: Indicates a preference for routing traffic to
+  endpoints that are on the same node as the client. In general, the
+  proxy should always route to a same-node endpoint if any is
+  available.
+
+We will add a new field to `discoveryv1.EndpointHints`:
+
+```golang
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+        ...
+
+	// forNodes indicates the node(s) this endpoint should be targeted by.
+	// +listType=atomic
+	ForNodes []string `json:"forNodes,omitempty" protobuf:"bytes,2,name=forNodes"`
+}
+```
+
+When updating EndpointSlices, if the EndpointSlice controller sees a
+service with `PreferSameNode` traffic distribution, then for each
+endpoint in the slice, it will add a `ForNodes` hint including the
```

Comment:
```suggestion
endpoint in the slice, it computes an internal `ForNodes` hint including the
```

If this isn't an internal detail of kube-controller-manager, I'm puzzled about why https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/ doesn't mention either hints or ForNodes.

Reply:
It doesn't mention `ForNodes` because that field is new in this KEP.

It doesn't mention hints because the "Topology information" subsection of that document apparently never got updated to explain the `TopologyAwareHints` feature... but it's not any more "internal" than anything else in EndpointSlice.